### PR TITLE
Add QgsCoordinateReferenceSystem::factor()

### LINKED ISSUE #41952: Expose more Proj's get_factors function through QgsCoordinateTransform or another API
**Feature description.**
I would love to use [Proj's get_factors](https://proj.org/development/reference/functions.html?highlight=proj_factors#c.proj_factors) function but it is not exposed by QGIS' Proj-ish QgsCoordinateTransform or any other API as I know of.

**Additional context**
I would like to release a rewrite of https://github.com/dtutic/ProjFactors but am stuck because of this. While it is possible to use the function via `pyproj`, you need to use a QGIS that is built using the same Proj version as the `pyproj` you end up importing or everything will crash (ref https://github.com/qgis/QGIS/issues/37289)

There might be more functions of Proj suitable for inclusion, I did not check for that.

<3

### PR DESCRIPTION
**Pull request by nyalldawson**
Calculates various cartographic properties, such as scale factors,
angular distortion and meridian convergence for the CRS at a
given geodetic point (in geographic coordinates).

Depending on the underlying projection values will be calculated
either numerically (default) or analytically. The function also
calculates the partial derivatives of the given coordinate.

Internally uses the proj library proj_factors API to calculate the factors.

Fixes #41952

### CODE REVIEW

File: tests/src/core/testqgscoordinatereferencesystem.cpp [hunk 1]
```diff
@@ -398,6 +400,87 @@ void TestQgsCoordinateReferenceSystem::fromProj4EPSG20936()
...
+  QGSCOMPARENEAR( factors.meridianParallelAngle(), 90.0, 0.01 );
+  QGSCOMPARENEAR( factors.meridianConvergence(), 0.0, 0.00000001 );
+  QGSCOMPARENEAR( factors.tissotSemimajor(), 1.0024419, 0.00000001 );
+  QGSCOMPARENEAR( factors.tissotSemiminor(), 1.0024419, 0.00000001 );
+  QGSCOMPARENEAR( factors.dxDlam(), 1.0, 0.00000001 );
+  QGSCOMPARENEAR( factors.dxDphi(), 0.0, 0.00000001 );
+  QGSCOMPARENEAR( factors.dyDlam(), 0.0, 0.00000001 );
+  QGSCOMPARENEAR( factors.dyDphi(), 1.0024419, 0.00000001 );
+
+  factors = QgsCoordinateReferenceSystem( QStringLiteral( "EPSG:3717" ) ).factors( QgsPoint( -120, 34 ) );
```
Comment (rouault) [inline tests/src/core/testqgscoordinatereferencesystem.cpp:468]: I would only keep that test, which is the one with the most non-trivial values. Other tests are mostly testing PROJ, not QGIS

Comment (nyalldawson) [reply tests/src/core/testqgscoordinatereferencesystem.cpp:408]: Done

**Review by rouault — APPROVED** [review]

**Comment by nyalldawson** [comment]
Thanks @rouault !

**Comment by kannes** [comment]
Thanks so much! I will probably build a plugin to make sampled rasters of these (https://github.com/qgis/QGIS/issues/41952#issue-819145528) at the end of May :)

**Pull request merged**